plugins/legion-hid/fu-legion-hid-device.c

```diff
@@ -0,0 +1,649 @@
+			}
+		}
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid_device_upgrade_verify(FuLegionHidDevice *self, guint8 id, GError **error)
+{
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(FuStructLegionHidUpgradeCmd) st_cmd = fu_struct_legion_hid_upgrade_cmd_new();
+	FuLegionHidUpgradeRetryHelper helper = {0};
+	guint8 status = 0;
+	guint8 content[] = {0x02, FU_LEGION_HID_UPGRADE_STEP_VERIFY, FU_LEGION_HID_CMD_CONSTANT_SN};
+
+	fu_struct_legion_hid_upgrade_cmd_set_length(st_cmd, sizeof(content) + 5);
+	fu_struct_legion_hid_upgrade_cmd_set_device_id(st_cmd, id);
+	if (!fu_struct_legion_hid_upgrade_cmd_set_data(st_cmd, content, sizeof(content), error))
+		return FALSE;
+	if (!fu_udev_device_write(FU_UDEV_DEVICE(self),
+				  st_cmd->buf->data,
+				  st_cmd->buf->len,
+				  FU_LEGION_HID_DEVICE_IO_TIMEOUT,
+				  FU_IO_CHANNEL_FLAG_NONE,
+				  error))
+		return FALSE;
+
+	res = g_byte_array_sized_new(FU_LEGION_HID_DEVICE_FW_REPORT_LENGTH);
+	helper.res = res;
+	helper.main_id = fu_struct_legion_hid_upgrade_cmd_get_main_id(st_cmd);
+	helper.sub_id = fu_struct_legion_hid_upgrade_cmd_get_sub_id(st_cmd);
+	helper.dev_id = id;
+	helper.step = FU_LEGION_HID_UPGRADE_STEP_VERIFY;
+	if (!fu_legion_hid_device_read_upgrade_response(self, &helper, error))
+		return FALSE;
+	status = res->data[9];
+	if (status != FU_LEGION_HID_RESPONSE_STATUS_OK) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "device report verify command failed with %u",
+			    status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static guint
+fu_legion_hid_device_get_version(FuLegionHidDevice *self, guint8 id, GError **error)
+{
+	guint version = 0;
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(FuStructLegionHidNormalCmd) st_cmd = fu_struct_legion_hid_normal_cmd_new();
+	FuLegionHidNormalRetryHelper helper = {0};
+	guint8 content[] = {FU_LEGION_HID_CMD_CONSTANT_SN};
+
+	fu_struct_legion_hid_normal_cmd_set_length(st_cmd, sizeof(content) + 4);
+	fu_struct_legion_hid_normal_cmd_set_main_id(st_cmd, 0x79);
+	fu_struct_legion_hid_normal_cmd_set_sub_id(st_cmd, 0x01);
+	fu_struct_legion_hid_normal_cmd_set_device_id(st_cmd, id);
+	if (!fu_struct_legion_hid_normal_cmd_set_data(st_cmd, content, sizeof(content), error))
+		return version;
+	if (!fu_udev_device_write(FU_UDEV_DEVICE(self),
+				  st_cmd->buf->data,
+				  st_cmd->buf->len,
+				  FU_LEGION_HID_DEVICE_IO_TIMEOUT,
+				  FU_IO_CHANNEL_FLAG_NONE,
+				  error))
+		return version;
+
+	res = g_byte_array_sized_new(FU_LEGION_HID_DEVICE_FW_REPORT_LENGTH);
+	helper.res = res;
+	helper.main_id = fu_struct_legion_hid_normal_cmd_get_main_id(st_cmd);
+	helper.sub_id = fu_struct_legion_hid_normal_cmd_get_sub_id(st_cmd);
+	helper.dev_id = id;
+	if (!fu_legion_hid_device_read_response(self, &helper, error))
+		return version;
+
+	version = fu_memread_uint32(res->data + 13, G_BIG_ENDIAN);
+	return version;
+}
+
+static gboolean
+fu_legion_hid_device_set_version(FuLegionHidDevice *device, GError **error)
```

Comment:
I think this should be called `fu_legion_hid_device_ensure_version` and it should be `FuLegionHidDevice *self` rather than `device`